OpenRA.Mods.Common/Activities/Rearm.cs

```diff
@@ -46,30 +43,32 @@ public override Activity Tick(Actor self)
 
 				needsReloading = true;
 
-				if (--ammoPoolsReloadTimes[pool] > 0)
-					continue;
+				Reload(self, pool);
+			}
+
+			return needsReloading ? this : NextActivity;
+		}
 
+		void Reload(Actor self, AmmoPool ammoPool)
+		{
+			if (!ammoPool.FullAmmo() && --ammoPool.RemainingTicks == 0)
```

Comment:
You already checked `FullAmmo()` before calling this method, so you don't need to recheck.